### DISCUSSION TITLE
packaging a parsing into a struct

### DISCUSSION THREAD

**Discussion by JaapWijnen** [pointfreeco/swift-parsing#239] [Q&A]
I recently wrote an xml parser using the parsing package. currently it lives in the form of a closure 
`let xmlParser: (Bool) -> AnyParser<Substring.UTF8View, XML>`
where the bool indicates if you want to print using indentation/pretty or just flattened xml, and the XML is my XML type to parse into. 
I want to package this into a struct so I can make the api a little bit cleaner:
- indicate what the bool is used for
- allow for passing strings, substrings or utf8views as input. 

I'm not entirely sure how to go about packaging it into a struct though since my parser uses a property (the boolean) to decide its behaviour. I currently think I can do two things: 

- Make a private computed property (this would maybe be recomputed every time? I'm not sure)
- Make a private parser property that is set on init

Both methods would require writing the whole parser type out for these properties (which is an incredibly long nested type), I could of course use .eraseToAnyParser() but I'm not sure if there's any cost involved with that?

Any advice on what way to go? 
My parser looks like this: 

```swift
ParsePrint {
    Optionally {
        xmlPrologParser // sub parser to parser the prolog
        Whitespace(.vertical).printing(indenting ? "\n".utf8 : "".utf8)
    }.map(Conversions.OptionalEmptyDictionary())
    containerTagParser(indenting ? 0 : nil) // sub parser to recursively go through the xml tree
    End()
}
```

so because of the use of the indenting variable I can't just do the following on init unfortunately
```swift
let parser = ParsePrint { /* the above parser builder */ }
```

**Comment by stephencelis** [comment]
> I could of course use .eraseToAnyParser() but I'm not sure if there's any cost involved with that?

This is probably the most ergonomic way of handling the problem right now. There is a slight cost at the point of erasure, but you should always benchmark before considering it a showstopper.

Swift 5.7 comes with some very powerful new generics features that will simplify this via `some ParserPrinter<Input, Output>`, but I assume you can't lock to 5.7 right now 😅

The future is bright, though!

**Reply by JaapWijnen** [reply]
Ha I'm lucky! I'm working on a few packages that depend on new 5.7 features so I might. But I wanted to publish it so other people can use it as well haha so I'll just keep it pinned to 5.5. Thanks for the advice! :) it's on the package index as swift-xml-parser if anyone is interested! If so let met know what you think! :)

**Reply by stephencelis** [reply]
If you want to mess around with it, we have a [`swift-5-7`](https://github.com/pointfreeco/swift-parsing/tree/swift-5-7) branch pushed. We're pretty focussed on a big TCA release right now, though, so we probably won't revisit it till closer to the GM.
